### PR TITLE
regex support for queries

### PR DESCRIPTION
**Pull request by anuchak**

### CODE REVIEW

File: src/include/common/expression_type.h [hunk 1]
```diff
@@ -114,6 +114,7 @@ const string ARRAY_EXTRACT_FUNC_NAME = "ARRAY_EXTRACT";
 const string CONCAT_FUNC_NAME = "CONCAT";
 const string CONTAINS_FUNC_NAME = "CONTAINS";
 const string ENDS_WITH_FUNC_NAME = "ENDS_WITH";
+const string REG_EXPR_FUNC_NAME = "REG_EXPR";
```
Comment (ray6080) [inline src/include/common/expression_type.h:117]: REG_EXPR -> RE_MATCH
REG_EXPR_FUNC_NAME -> RE_MATCH_FUNC_NAME

File: src/include/function/string/operations/reg_expr_operation.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+struct RegExpr {
+    static inline void operation(ku_string_t& left, ku_string_t& right, uint8_t& result) {
+        result = RE2::FullMatch(left.getAsString(), right.getAsString());
+    }
+};
+
+} // namespace operation
+} // namespace function
+} // namespace kuzu
```
Comment (ray6080) [inline src/include/function/string/operations/reg_expr_operation.h:22]: New line here.

File: src/include/function/string/operations/reg_expr_operation.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+#include "re2.h"
+
+using namespace kuzu::common;
+using namespace kuzu::regex;
+
+namespace kuzu {
+namespace function {
+namespace operation {
+
+struct RegExpr {
```
Comment (ray6080) [inline src/include/function/string/operations/reg_expr_operation.h:14]: The general suggestion is to not call the operation/function as RegExpr, instead be more specific, this is a regular expression full match function. let's rename them follow the name "RegExprMatch" or "REMatch".

File: src/include/function/string/vector_string_operations.h [hunk 4]
```diff
@@ -63,6 +63,10 @@ struct EndsWithVectorOperation : public VectorStringOperations {
     static vector<unique_ptr<VectorOperationDefinition>> getDefinitions();
 };
 
+struct RegExprVectorOperation : public VectorStringOperations {
```
Comment (ray6080) [inline src/include/function/string/vector_string_operations.h:66]: Ditto. RegExprVectorOperation -> RegExprMatchVectorOperation or REMatchVectorOperation

File: third_party/CMakeLists.txt [hunk 5]
```diff
@@ -2,3 +2,4 @@ add_subdirectory(antlr4_runtime)
 add_subdirectory(antlr4_cypher)
 add_subdirectory(utf8proc)
 add_subdirectory(pybind11)
+add_subdirectory(re2)
```
Comment (ray6080) [inline third_party/CMakeLists.txt:5]: New line here.

File: test/test_files/tinysnb/function/string.test [hunk 6]
```diff
@@ -144,6 +144,22 @@ False
 ---- 1
 Carol
 
+-NAME RegExprString
```
Comment (ray6080) [inline test/test_files/tinysnb/function/string.test:147]: RegExprString -> RegExprMatchString

Same for following.

File: third_party/re2/bitstate.cpp [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    bool longest = kind != kFirstMatch;
+    if (!b.Search(text, context, anchored, longest, match, nmatch))
+        return false;
+    if (kind == kFullMatch && EndPtr(match[0]) != EndPtr(text))
+        return false;
+    return true;
+}
+
+} // namespace regex
+} // namespace kuzu
```
Comment (ray6080) [inline third_party/re2/bitstate.cpp:374]: New line here.

File: test/test_files/tinysnb/function/string.test [hunk 8]
```diff
@@ -144,6 +144,22 @@ False
...
+-QUERY MATCH (a:person) UNWIND a.usedNames as x WITH x AS name WHERE name =~ '^A.*$' RETURN name
+---- 2
+Ad
+Aida
+
+-NAME RegExprUTF8String
+-QUERY MATCH (m:movies) WHERE m.name =~ '.*â.*' RETURN m.name
+---- 1
+Sóló cón tu párejâ
+
```
Comment (ray6080) [inline test/test_files/tinysnb/function/string.test:162]: I think it's better to add a bit more expressions for tests here. e.g., the one with escapes.

**Review by ray6080 — APPROVED** [review]
LGTM. Thanks @anuchak

File: src/include/function/string/operations/reg_expr_operation.h (hunk 2 above)
Comment (anuchak) [reply src/include/function/string/operations/reg_expr_operation.h]: done

File: src/include/function/string/operations/reg_expr_operation.h (hunk 3 above)
Comment (anuchak) [reply src/include/function/string/operations/reg_expr_operation.h]: done

File: src/include/function/string/vector_string_operations.h (hunk 4 above)
Comment (anuchak) [reply src/include/function/string/vector_string_operations.h]: done

File: third_party/CMakeLists.txt (hunk 5 above)
Comment (anuchak) [reply third_party/CMakeLists.txt]: done

File: test/test_files/tinysnb/function/string.test (hunk 6 above)
Comment (anuchak) [reply test/test_files/tinysnb/function/string.test]: changed

File: third_party/re2/bitstate.cpp (hunk 7 above)
Comment (anuchak) [reply third_party/re2/bitstate.cpp]: added

File: test/test_files/tinysnb/function/string.test (hunk 8 above)
Comment (anuchak) [reply test/test_files/tinysnb/function/string.test:162]: added 3 more tests with escape sequence regex

**Pull request merged**